OpenRA.Game/Server/Server.cs

```diff
@@ -625,11 +678,111 @@ void DispatchOrdersToClient(Connection c, int client, int frame, byte[] data)
 			}
 		}
 
+		// Check if only one side remains
+		bool HasRemainingConflict()
+		{
+			var lastTeam = -1;
+			var remainingPlayers = gameInfo.Players.Where(p => p.Outcome == WinState.Undefined);
+			foreach (var player in remainingPlayers)
+			{
+				if (lastTeam >= 0 && (player.Team != lastTeam || player.Team == 0))
+					return true;
+
+				lastTeam = player.Team;
+			}
+
+			return false;
+		}
+
+		void SetRemainingPlayersAsWinners()
+		{
+			var now = DateTime.UtcNow;
+			var remainingPlayers = gameInfo.Players.Where(p => p.Outcome == WinState.Undefined);
+			foreach (var winner in remainingPlayers)
+			{
+				winner.Outcome = WinState.Won;
+				winner.OutcomeTimestampUtc = now;
+			}
+		}
+
+		void SetPlayerDefeat(int client)
+		{
+			var defeatedPlayer = gameInfo.Players.FirstOrDefault(p => p.ClientIndex == client);
+			if (defeatedPlayer == null || defeatedPlayer.Outcome != WinState.Undefined)
+				return;
+
+			defeatedPlayer.Outcome = WinState.Lost;
+			defeatedPlayer.OutcomeTimestampUtc = DateTime.UtcNow;
+
+			if (!HasRemainingConflict())
+				SetRemainingPlayersAsWinners();
+		}
+
+		bool OutOfSync(int frame)
+		{
+			Log.Write("server", "Out of sync detected at frame {0}, cancel replay recording", frame);
+
+			// Make sure the written file is not valid
+			recorder.Metadata = null;
+
+			recorder.Dispose();
+
+			// Stop the recording
+			recorder = null;
+
+			return false;
+		}
+
+		Dictionary<int, byte[]> syncForFrame = new Dictionary<int, byte[]>();
+		int lastFrameDefeatState;
```

Comment:
Naming nit: `lastFrameDefeatState` implies that this should hold the defeat state for the last frame - so can we change this to `lastDefeatStateFrame`?

Reply:
Yes; also more consistent with `lastDefeatState`. Renamed.